src/lib/optimizer/strategy/predicate_merge_rule.cpp

```diff
@@ -0,0 +1,105 @@
+#include "predicate_merge_rule.hpp"
+
+#include "expression/expression_functional.hpp"
+#include "expression/logical_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/union_node.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+PredicateMergeRule::PredicateMergeRule(const size_t optimization_threshold)
+    : _optimization_threshold(optimization_threshold) {}
+
+void PredicateMergeRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  Assert(root->type == LQPNodeType::Root, "PredicateMergeRule needs root to hold onto");
+
+  visit_lqp(root, [&](const auto& node) {
+    size_t lqp_complexity = 0;
+    visit_lqp(node, [&](const auto &sub_node) {
+      switch (sub_node->type) {
+        case LQPNodeType::Predicate:
+          return LQPVisitation::VisitInputs;
+
+        case LQPNodeType::Union:
+          lqp_complexity++;
+          return LQPVisitation::VisitInputs;
+
+        default:
+          return LQPVisitation::DoNotVisitInputs;
+      }
+    });
+
+    // Simple heuristic: The PredicateMergeRule is more likely to improve the performance for complex LQPs with many
+    // UNIONs.
+    if (lqp_complexity >= _optimization_threshold) {
+      _merge_subplan(node, std::nullopt);
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+}
+
+/**
+ * Merge an LQP that only consists of PredicateNodes and UnionNodes into a single PredicateNode. The
+ * subsequent_expression parameter passes the translated expressions to the translation of its children nodes, which
+ * enables to add the translated expression of child node before its parent node to the output expression.
+ */
+std::shared_ptr<AbstractExpression> PredicateMergeRule::_merge_subplan(
+    const std::shared_ptr<AbstractLQPNode>& begin,
+    const std::optional<const std::shared_ptr<AbstractExpression>>& subsequent_expression) const {
```

Comment:
This function is loosely based on `lqp_subplan_to_boolean_expression` in `lqp_utils`. The major difference it that the LQP is traversed and changed at the same time here.